node-websocket-chat/README.md

```diff
@@ -0,0 +1,202 @@
+* `writer_user_message` writes a `Chat::UserMessage` to the `User Messages` topic.
+* `subscribe_user_messages` takes a callback that will be invoked whenever a user message is received.  This code enables reliable data transfer which is off by default for a subscriber.
+
+The [server.js](server.js) file is based on the server code for https://karlhadwen.medium.com/node-js-websocket-tutorial-real-time-chat-room-using-multiple-clients-44a8e26a953e.
+The code now contains logic for handling the listening port of the web server so multiple servers can run on the same host and the necessary logic for introducing the OpenDDS wrapper and initializing OpenDDS.
+
+The major change is the code that processes messages received via websocket.
+In the original server code, a message that was received on a websocket connect was sent to all of the connection except the one that originally sent the message.
+This code has been broken into two separate pieces:  one for receiving and one for sending.
+The code that handles messages from the websockets writes the received message to the `User Messages` topic.
+When initializing OpenDDS, the code installs a callback that sends a message received on the `User Messages` topic to all of the websocket clients that are connected.
+Together, these changes result in the following flow for a message:
+
+1. The client sends a message.
+2. The server receives the message via websocket and writes it to the `User Messages` topic.
+3. The server receives the message via the `User Messages` topic and writes it to all connected websocket clients.
+4. The client receives the message from the websocket and displays it.
+
+Thus, there is two levels of fan-out.
+First, a message is distributed to all of the server instances.
+Then, it is distributed to all of the clients of each server instance.
+
+## Building and Running
+
+1. [Build OpenDDS](https://opendds.readthedocs.io/en/latest-release/building/index.html).  Let `DDS_ROOT` be the directory containing OpenDDS.
+
+        source "${DDS_ROOT}/setenv.sh"
+
+2. Clone this repo.
+
+        git clone https://github.com/OpenDDS/DeveloperResources.git
+
+3. Enter the directory for this article.
+
+        cd DeveloperResources/node-websocket-chat
+    
+4. Run mwc.
+
+        mwc.pl -type gnuace
+    
+5. Run make
+
+        make
+
+6. Export `CHAT_ROOT` for library loading purposes.
+
+        export CHAT_ROOT="${PWD}"
+
+7. Update `LD_LIBRARY_PATH`
+
+        export LD_LIBRARY_PATH="${CHAT_ROOT}/idl"
+    
+8. Install node modules.
+
+        npm install
+    
+9. Start the servers in the background 
+
+        node server.js --port 7000 -DCPSConfigFile rtps.ini &
+        node server.js --port 7001 -DCPSConfigFile rtps.ini &
+    
+10. Open `index.html` in a web browser.  Use two tabs or windows to communicate with each server.
+
+    1. In the first tab/window, connect to the first server by using `7000` for the port and clicking "Connect".
```

Comment:
I went through this and it worked fine. I know this is supposed to be bare bones, but it'd be nice if there was something that indicated if it was connected or not.